src/Tables.jl

```diff
@@ -148,34 +216,89 @@ end
 """
 allocatecolumn(T, len) = Vector{T}(undef, len)
 
-@inline function allocatecolumns(::Type{NT}, len) where {NT <: NamedTuple{names}} where {names}
+@inline function allocatecolumns(::Schema{names, types}, len) where {names, types}
     if @generated
-        vals = Tuple(:(allocatecolumn($(fieldtype(NT, i)), len)) for i = 1:fieldcount(NT))
+        vals = Tuple(:(allocatecolumn($(fieldtype(types, i)), len)) for i = 1:fieldcount(types))
         return :(NamedTuple{names}(($(vals...),)))
     else
-        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(NT, i), len) for i = 1:fieldcount(NT)))
+        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(types, i), len) for i = 1:fieldcount(types)))
     end
 end
 
+haslength(T) = T === Base.HasLength() || T === Base.HasShape{1}()
+
 # add! will push! or setindex! a value depending on if the row-iterator HasLength or not
-@inline add!(val, col::Int, nm::Symbol, ::Base.HasLength, nt, row) = setindex!(nt[col], val, row)
+@inline add!(val, col::Int, nm::Symbol, ::Union{Base.HasLength, Base.HasShape{1}}, nt, row) = setindex!(nt[col], val, row)
 @inline add!(val, col::Int, nm::Symbol, T, nt, row) = push!(nt[col], val)
 
-@inline function buildcolumns(sch, rowitr::T) where {T}
+@inline function buildcolumns(schema, rowitr::T) where {T}
     L = Base.IteratorSize(T)
-    len = L == Base.HasLength() ? length(rowitr) : 0
-    nt = allocatecolumns(sch, len)
+    len = haslength(L) ? length(rowitr) : 0
+    nt = allocatecolumns(schema, len)
     for (i, row) in enumerate(rowitr)
-        unroll(add!, sch, row, L, nt, i)
+        eachcolumn(add!, schema, row, L, nt, i)
     end
     return nt
 end
 
+@inline function add_or_widen!(val::T, col::Int, nm::Symbol, L, columns, row, allocate, len) where {T}
+    if !allocate
+        @inbounds columns[col] = add_or_widen!(columns[col], val, L, row)
+    else
+        @inbounds columns[col] = add_or_widen!(allocatecolumn(T, len), val, L, row)
+    end
+    return
+end
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, ::Union{Base.HasLength, Base.HasShape{1}}, row) where {T, S}
+    if S === T || val isa T
+        @inbounds dest[row] = val
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), length(dest))
+        copyto!(new, 1, dest, 1, row - 1)
+        @inbounds new[row] = val
+        return new
+    end
+end
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, ::Base.SizeUnknown, row) where {T, S}
+    if S === T || val isa T
+        push!(dest, val)
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), length(dest))
+        copyto!(new, dest)
+        push!(new, el)
+        return new
+    end
+end
+
+# when Tables.types(x) === nothing
+function buildcolumns(sch::Schema{names, nothing}, rowitr::T) where {names, T}
+    state = iterate(rowitr)
+    state === nothing && return NamedTuple{names}(Tuple(Missing[] for _ in names))
```

Comment:
IndexedTables uses inference in this case to guess the type, though returning `Missing` type like you do here is probably more principled. We should be consistent, but maybe it's the IndexedTables side that needs to change?

Reply:
Why `Missing[]` rather than `Union{}`? Anyway it would make sense to use inference just like `map`.

Reply:
There is no "inference" here, we're not applying a function, we're just getting an empty iterator. I considered `Union{}`, but it felt to Base-collect-y to me: here we're talking tables in the data world, where `Missing` seems a more natural default, like an all `NULL` column in a database.

Reply:
What I'm worried about with `Missing` is that if you later combine the resulting table with another (non-empty) one, you'll get a `Union{T,Missing}` column, which wouldn't really be appropriate.

Reply:
Hmmmm, that's a good point; that certainly makes a case for `Union{}` instead.

Reply:
You possibly want an iterator that has eltype `Union{}` and zero elements. You could e.g. create an `struct EmptyVector <: AbstractVector{Union{}}; end` (or something similar) for this, which will probably behave nicely under `vcat` and so-on.

Reply:
Sorry, I probably wasn't clear, I mean the columns should be like a `Vector{Union{}}()` instead of a `Vector{Missing}()`.

But it's also ever-so-slightly wasteful to allocate `Vector` here for containers you know are empty (this is only important in the case you have many, many small or empty tables... sorry, I'm the StaticArrays guy so I tend to think of these cases and try to minimize overhead for small containers...)